cranelift/codegen/src/isa/s390x/inst/emit.rs

```diff
@@ -1654,52 +1654,94 @@ impl MachInstEmit for Inst {
                 put(sink, &enc_ril_a(opcode, rd.to_reg(), imm.bits));
             }
 
-            &Inst::SMulWide { rn, rm } => {
+            &Inst::SMulWide { rd, rn, rm } => {
                 let rn = allocs.next(rn);
                 let rm = allocs.next(rm);
+                let rd1 = allocs.next_writable(rd.hi);
+                let rd2 = allocs.next_writable(rd.lo);
+                debug_assert_eq!(rd1, writable_gpr(2));
+                debug_assert_eq!(rd2, writable_gpr(3));
```

Comment:
There's actually no need now to assert we get the specific r2/r3 pair here.  The emit code might as well accept any valid register pair, i.e. only assert that we have an even/odd GPR pair, and that the tied input (if any) matches the output.

Then the only place where r2/r3 are still hardcoded is in `s390x_get_operands`.

Reply:
That's much cleaner, thanks!